scripts/vci-directory-auditor/src/index.ts

```diff
@@ -1,172 +1,255 @@
+            if (!acaoHeader) {
+                issuerLogInfo.warnings?.push("Issuer key endpoint does not contain a CORS 'access-control-allow-origin' header");
+            } else if (acaoHeader !== '*' && acaoHeader !== requestedOrigin) {
+                issuerLogInfo.warnings?.push(`Issuer key endpoint's CORS 'access-control-allow-origin' header ${acaoHeader} does not match the requested origin`);
             }
             const keySet = JSON.parse(response.body) as KeySet;
             if (!keySet) {
-                throw "Failed to parse JSON KeySet schema";
+                issuerLogInfo.errors?.push("Failed to parse JSON KeySet schema");
             }
+            // TODO: try to import keys
             issuerLogInfo.keys = keySet.keys;
         } catch (err) {
             issuerLogInfo.errors?.push((err as Error).toString());
         }
-        return issuerLogInfo;        
-    });
+        try {
+            issuerLogInfo.tlsDetails = getDefaultTlsDetails(new Url(issuer.iss).hostname);
+            if (issuerLogInfo.tlsDetails) {
+                // we report TLS issues as warnings
+                auditTlsDetails(issuerLogInfo.tlsDetails).map(a => issuerLogInfo.warnings?.push(a));
+            }
+        } catch (err) {
+            // report TLS issues as warnings
+            issuerLogInfo.warnings?.push((err as Error).toString());
+        }
+        for (const key of issuerLogInfo.keys) {
+            // the issuer has a card revocation list (CRL); fetch it
+            if (key.crlVersion) {
+                try {
+                    const crlURL = `${issuer.iss}/.well-known/crl/${key.kid}.json`;
+                    if (verbose) console.log("fetching CRL at " + crlURL);
+                    const response = await got(crlURL, { timeout:5000 });
+                    if (!response) {
+                        throw "Can't reach CRL at " + crlURL;
+                    }
+                    const crl = JSON.parse(response.body) as CRL;
+                    if (!crl) {
+                        throw "Can't parse CRL at " + crlURL;
+                    }
+                    issuerLogInfo.crls?.push(crl);
+                } catch (err) {
+                    issuerLogInfo.errors?.push((err as Error).toString());
+                }
+            }
+        }
+        issuerLogInfoArray.push(issuerLogInfo);
+        process.stdout.write("."); // print progress marker (without newline, unlike console.log)
+    }
+    console.log("done"); // also puts a newline after the progress markers
 
-    const directoryLog: DirectorySnapshot = {
-        directory: VCI_ISSUERS_DIR_URL,
+    const directoryLog: DirectoryLog = {
+        directory: "https://raw.githubusercontent.com/the-commons-project/vci-directory/main/vci-issuers.json", // directoryUrl, TODO
         time: date.format(currentTime, 'YYYY-MM-DD HH:mm:ss'),
-        issuerInfo: []    
-    }
-    try {
-        directoryLog.issuerInfo = await Promise.all(issuerLogInfo);
```

Comment:
Curious - were there any issues with this approach? In testing, it seems to speed up the runs dramatically (not that it matters much now).

Reply:
I ran into issues when fetching the keys through a map. I think the `got` package couldn't handle all the connections, and they started to timeout (probably started, queued, then performed by platform lib). Anyway, I switched to a predictable iteration flow, and the timeout issues was resolved. There are probably ways to improve the efficiency of this script, if needed.